bionic/protocols.py

```diff
@@ -599,6 +599,69 @@ def read(self, path):
         return single_element(path.iterdir())
 
 
+# This protocol allows us to hash sets deterministically.
+# It would be preferable if PicklableProtocol could install a special handler for sets,
+# but unfortunately pickle's override mechanisms (dispatch tables and reduce_override)
+# don't work for built-in types like set -- at least not in the C implementation of
+# pickle. This custom protocol will work for set objects, but not for any other objects
+# that happen to contain sets.
+class PicklableSetProtocol(BaseProtocol):
+    """
+    Decorator indicating that an entity's value will be a set or frozenset whose
+    contents are picklable.
+
+    Sets need special handling because their iteration order is non-deterministic, so
+    the result of pickling them is also non-deterministic, which causes problems when
+    the serialized file is hashed. To resolve this, this protocol attempts to sort
+    the set before pickling it. (If the set is not sortable, we emit a warning and
+    pickle it anyway.)
+
+    Parameters:
+        pickle_protocol_version: int (default: 4)
+            The pickle serialization protocol to use.
+    """
+
+    def __init__(self, pickle_protocol_version=4):
+        super(PicklableSetProtocol, self).__init__()
+        self._pickle_protocol_version = pickle_protocol_version
+
+    def get_fixed_file_extension(self):
+        return "setpkl"
+
+    def validate(self, value):
+        assert isinstance(value, (set, frozenset))
+
+    def write(self, value, path):
+        original_type = type(value)
+        try:
+            sorted_contents = sorted(value)
+        except TypeError as e:
+            message = f"""
```

Comment:
If sorting on the set fails, another alternative is to convert all elements to bytes using `pickle.dumps` and sort the bytes. This would require some extra pickling and unpickling logic, but we can be sure that any picklable set can use `PicklableSetProtocol`.

Reply:
That's a great idea. I can think of two ways to do this (that don't involve redundant pickling):

1. Pickle all the elements, sort the resulting collection of bytes objects, then pickle that sorted collection.
2. Create a directory and pickle each element to a file in the directory. Then somehow sort the files (by hash or by contents) at some point during saving or loading.

1 is simpler but requires making an in-memory copy of the whole thing. WDYT?

Reply:
I think the extra space in 1 is okay since we are not going to keep the in-memory copy around for long. My hesitation with 2 is that it might get tricky creating digest hash for these files (since we'll have to somehow sort them).

Reply:
I'm always reluctant to make a change that could potentially double the peak memory usage of a flow, but I agree that 2 is tricky. I'll try and see if there's a simple-ish way to do 2.

Reply:
I tried number 2 and it didn't seem too bad. Let me know what you think.

Reply:
I didn't realize that our tokenization function accepts a folder too. This was easier than I expected. But this will also convert `frozenset` to `set`. I think we should pickle and store the type of the set too. What do you think?

Reply:
Good catch! That should have a test!

(Actually I originally considered having sets be deserialized as frozensets, because no one should be mutating them anyway, but I thought it was also potentially confusing. Could still be convinced though.)

Reply:
I fixed this and parametrized one of the tests to check frozensets as well.